Signaling/app/src/main/java/com/tokbox/android/tutorials/signaling/MainActivity.java

```diff
@@ -0,0 +1,362 @@
+            mMessageHistory = new ChatMessageAdapter(this);
+            mMessageHistoryListView.setAdapter(mMessageHistory);
+
+            // Attach handlers to UI
+            mSendButton.setOnClickListener(this);
+
+            // if there is no server URL set
+            if (OpenTokConfig.CHAT_SERVER_URL == null) {
+                // use hard coded session values
+                if (OpenTokConfig.areHardCodedConfigsValid()) {
+                    initializeSession(OpenTokConfig.API_KEY, OpenTokConfig.SESSION_ID, OpenTokConfig.TOKEN);
+                    initializePublisher();
+                } else {
+                    showConfigError("Configuration Error", OpenTokConfig.hardCodedConfigErrorMessage);
+                }
+            } else {
+                // otherwise initialize WebServiceCoordinator and kick off request for session data
+                // session initialization occurs once data is returned, in onSessionConnectionDataReady
+                if (OpenTokConfig.isWebServerConfigUrlValid()) {
+                    mWebServiceCoordinator = new WebServiceCoordinator(this, this);
+                    mWebServiceCoordinator.fetchSessionConnectionData(OpenTokConfig.SESSION_INFO_ENDPOINT);
+                } else {
+                    showConfigError("Configuration Error", OpenTokConfig.webServerConfigErrorMessage);
+                }
+            }
+        } else {
+            EasyPermissions.requestPermissions(this, getString(R.string.rationale_video_app), RC_VIDEO_APP_PERM, perms);
+        }
+    }
+
+    @Override
+    public boolean onOptionsItemSelected(MenuItem item) {
+        // Handle action bar item clicks here. The action bar will
+        // automatically handle clicks on the Home/Up button, so long
+        // as you specify a parent activity in AndroidManifest.xml.
+        int id = item.getItemId();
+
+        //noinspection SimplifiableIfStatement
+        if (id == R.id.action_settings) {
+            return true;
+        }
+
+        return super.onOptionsItemSelected(item);
+    }
+
+    private void initializeSession(String apiKey, String sessionId, String token) {
+        Log.d(LOG_TAG, "Initializing Session");
+        Log.d(LOG_TAG, "API Key: " + apiKey);
+        Log.d(LOG_TAG, "session ID: " + sessionId);
+        Log.d(LOG_TAG, "token: " + token);
+
+        mSession = new Session.Builder(this, apiKey, sessionId).build();
+        mSession.setSessionListener(this);
+        mSession.setSignalListener(this);
+        mSession.connect(token);
+    }
+
+    private void initializePublisher() {
+        mPublisher = new Publisher.Builder(this).build();
+        mPublisher.setPublisherListener(this);
+        mPublisher.getRenderer().setStyle(BaseVideoRenderer.STYLE_VIDEO_SCALE,
+                BaseVideoRenderer.STYLE_VIDEO_FILL);
+        mPublisherViewContainer.addView(mPublisher.getView());
+    }
+
+    private void sendMessage() {
+        disableMessageViews();
```

Comment:
I am not sure if we need to block the edit TextMessage from the activity while we are sending the signal